File: C/fors.c
#include <string.h>
#include <math.h>
#include "params.h"
#include "adrs.h"
#include "shake.h"
#include "wots.h"
#include <stdio.h>

// algorithm 14
void fors_skGen(Parameters *prm, const uint8_t *sk_seed, const uint8_t *pk_seed, ADRS adrs, uint64_t idx, uint8_t *buffer)
{
    ADRS sk_adrs;
    sk_adrs = adrs;
    setTypeAndClear(&sk_adrs, prm->FORS_PRF);
    setKeyPairAddress(&sk_adrs, getKeyPairAddress(&adrs));
    setTreeIndex(&sk_adrs, idx);
    PRF(prm, pk_seed, sk_seed, &sk_adrs, buffer);
}

// Algorithm 15 (Computes the root of a Merkle subtree of FORS public values)
void fors_node(Parameters *prm, const uint8_t *sk_seed, uint64_t i, uint64_t z, const uint8_t *pk_seed, ADRS adrs, uint8_t *buffer)
{
    uint8_t node[prm->n];

    if (z == 0) {
        uint8_t sk[prm->n];
        fors_skGen(prm, sk_seed, pk_seed, adrs, i, sk);
        setTreeHeight(&adrs, 0);
        setTreeIndex(&adrs, i);
        F(prm, pk_seed, &adrs, sk, node);
    } else {
        uint8_t lnode[prm->n];
        uint8_t rnode[prm->n];
        uint8_t combined[2 * prm->n];

        fors_node(prm, sk_seed, 2 * i,     z - 1, pk_seed, adrs, lnode);
        fors_node(prm, sk_seed, 2 * i + 1, z - 1, pk_seed, adrs, rnode);

        setTreeHeight(&adrs, z);
        setTreeIndex(&adrs, i);

        memcpy(combined, lnode, prm->n);
        memcpy(combined + prm->n, rnode, prm->n);

        H(prm, pk_seed, &adrs, combined, node);
    }
    memcpy(buffer, node, prm->n);
}

// Algorithm 16 (Generates a FORS signature)
void fors_sign(Parameters *prm, const uint8_t *md, const uint8_t *sk_seed, const uint8_t *pk_seed, ADRS adrs, uint8_t *buffer)
{
    uint32_t sig_len = prm->n + prm->a * prm->n;
    uint8_t sig_fors[sig_len * prm->k];
    uint8_t indices[prm->k];
    uint8_t auth[prm->a * prm->n];
    base_2b(md, prm->a, prm->k, indices);

    for (uint32_t i = 0; i < prm->k; i++) {
        fors_skGen(prm, sk_seed, pk_seed, adrs, i * (uint64_t) pow(2, prm->a) + indices[i], sig_fors + i * sig_len);

        for (uint32_t j = 0; j < prm->a; j++) {
            uint64_t s = (uint64_t) floor(indices[i] / pow(2, j)) ^ 1;
            fors_node(prm, sk_seed, i * (uint64_t) pow(2, prm->a - j) + s, j, pk_seed, adrs, auth + j * prm->n);
        }
        memcpy(sig_fors + i * sig_len + prm->n, auth, prm->a * prm->n);
    }
    memcpy(buffer, sig_fors, sig_len * prm->k);
}

// Algorithm 17 (Computes a FORS public key from a FORS signature)
void fors_pkFromSig(Parameters *prm, uint8_t *sig_fors, const uint8_t *md, const uint8_t *pk_seed, ADRS adrs, uint8_t *buffer)
{
    uint32_t sig_len = prm->n + prm->a * prm->n;
    uint8_t indices[prm->k];
    base_2b(md, prm->a, prm->k, indices);

    uint8_t sk[prm->n];
    uint8_t root[prm->k * prm->n];
    uint8_t node_0[prm->n];
    uint8_t node_1[prm->n];
    uint8_t auth[prm->a * prm->n];
    uint8_t combined[2 * prm->n];

    for (uint32_t i = 0; i < prm->k; i++) {
        memcpy(sk, sig_fors + i * sig_len, prm->n);
        setTreeHeight(&adrs, 0);
        setTreeIndex(&adrs, i * (uint64_t) pow(2, prm->a) + indices[i]);
        F(prm, pk_seed, &adrs, sk, node_0);
        memcpy(auth, sig_fors + i * sig_len + prm->n, prm->a * prm->n);

        for (uint32_t j = 0; j < prm->a; j++) {

            setTreeHeight(&adrs, j + 1);
            if ((indices[i] / (uint64_t) pow(2, j)) % 2 == 0) {
                setTreeIndex(&adrs, getTreeIndex(&adrs) / 2);
                memcpy(combined, node_0, prm->n);
                memcpy(combined + prm->n, auth + j * prm->n, prm->n);
                H(prm, pk_seed, &adrs, combined, node_1);
            } else {
                setTreeIndex(&adrs, (getTreeIndex(&adrs) - 1) / 2);
                memcpy(combined, auth + j * prm->n, prm->n);
                memcpy(combined + prm->n, node_0, prm->n);
                H(prm, pk_seed, &adrs, combined, node_1);
            }
            memcpy(node_0, node_1, prm->n);
        }
        memcpy(root + i * prm->n, node_0, prm->n);
    }
    ADRS forspkadrs;
    forspkadrs = adrs;
    setTypeAndClear(&forspkadrs, prm->FORS_ROOTS);
    setKeyPairAddress(&forspkadrs, getKeyPairAddress(&adrs));
    Tlen(prm, pk_seed, &forspkadrs, root, prm->k * prm->n, buffer);
}

File: python/fors.py
from copy import deepcopy

import params
from adrs import ADRS
from shake import F, H, PRF, Tlen
from wots import base_2b


# algorithm 14
def fors_skGen(sk_seed: bytes, pk_seed: bytes, adrs: ADRS, idx: int) -> bytes:
    """
    Generates a FORS private-key value

    Params:
        sk_seed     secret seed
        pk_seed     public seed
        ADRS        address
        idx         secret key index
    Returns:
        n-byte FORS private key
    """
    sk_adrs = deepcopy(adrs)
    sk_adrs.setTypeAndClear(params.prm.FORS_PRF)
    sk_adrs.setKeyPairAddress(adrs.getKeyPairAddress())
    sk_adrs.setTreeIndex(idx)
    return PRF(pk_seed, sk_adrs, sk_seed)


#Algorithmus 15 (Computes the root of a Merkle subtree of FORS public values)
def fors_node(SK_seed: bytes, i: int, z: int, PK_seed: bytes, ADRS) -> bytes:    #Input: Secret seed SK.seed, target node index 𝑖, target node height 𝑧, public seed PK.seed, address ADRS

    if z == 0:
        sk = fors_skGen(SK_seed, PK_seed, ADRS, i)
        ADRS.setTreeHeight(0)
        ADRS.setTreeIndex(i)
        node = F(PK_seed, ADRS, sk)
    else:
        lnode = fors_node(SK_seed, 2 * i, z - 1, PK_seed, ADRS)
        rnode = fors_node(SK_seed, 2 * i + 1, z - 1, PK_seed, ADRS)
        ADRS.setTreeHeight(z)
        ADRS.setTreeIndex(i)

        concatenated_nodes = lnode + rnode
        node = H(PK_seed, ADRS, concatenated_nodes)

    return node                                 #Output: 𝑛-byte root 𝑛𝑜𝑑𝑒

#Algorithmus 16 (Generates a FORS signature)
def fors_sign(md: bytes, SK_seed: bytes, PK_seed: bytes, ADRS: ADRS) -> bytes:      #Input: Message digest 𝑚𝑑, secret seed SK.seed, address ADRS, public seed PK.seed

    SIG_FORS = b""
    indices = base_2b(md, params.prm.a, params.prm.k)                 # Compute indices using base_2b function

    for i in range(params.prm.k):
        SIG_FORS += fors_skGen(SK_seed, PK_seed, ADRS, i * 2 ** params.prm.a + indices[i])

        AUTH = b""
        for j in range(params.prm.a):
            s = (indices[i] >> j) ^ 1
            AUTH += fors_node(SK_seed, i * 2**(params.prm.a - j) + s, j, PK_seed, ADRS)
        SIG_FORS += AUTH

    return SIG_FORS                             # Output: FORS signature SIG𝐹 𝑂𝑅𝑆


#Algorithmus 17 (Computes a FORS public key from a FORS signature)
def fors_pkFromSig(SIG_FORS: bytes, md: bytes, PK_seed: bytes, ADRS: ADRS) -> bytes:

    indices = base_2b(md, params.prm.a, params.prm.k)
    root = b""

    for i in range(params.prm.k):
        sk = SIG_FORS[i * (params.prm.a + 1) * params.prm.n:(i * (params.prm.a + 1) + 1) * params.prm.n]
        ADRS.setTreeHeight(0)
        ADRS.setTreeIndex(i * 2 ** params.prm.a + indices[i])
        node_0 = F(PK_seed, ADRS, sk)

        auth = SIG_FORS[(i * (params.prm.a + 1) + 1) * params.prm.n:(i + 1) * (params.prm.a + 1) * params.prm.n]
        for j in range(params.prm.a):
            auth_j = auth[j * params.prm.n:(j + 1) * params.prm.n]
            ADRS.setTreeHeight(j + 1)
            if indices[i] // 2**j % 2 == 0:
                ADRS.setTreeIndex(ADRS.getTreeIndex() // 2)
                node_1 = H(PK_seed, ADRS, node_0 + auth_j);
            else:
                ADRS.setTreeIndex((ADRS.getTreeIndex() - 1) // 2)
                node_1 = H(PK_seed, ADRS, auth_j + node_0);

            node_0 = node_1
        root += node_0

    forspkADRS = deepcopy(ADRS)
    forspkADRS.setTypeAndClear(params.prm.FORS_ROOTS)
    forspkADRS.setKeyPairAddress(ADRS.getKeyPairAddress())

    pk = Tlen(PK_seed, forspkADRS, root)
    return pk
